tests/msc2403_test.go

```diff
@@ -0,0 +1,241 @@
+	knockingBetweenTwoUsersTest(t, roomIDTwo, alice, charlie, true)
+}
+
+func knockingBetweenTwoUsersTest(t *testing.T, roomID string, inRoomUser, knockingUser *client.CSAPI, federation bool) {
+	knockingUserUserID := knockingUser.UserID
+
+	t.Run("Knocking on a room with a join rule other than 'knock' should fail", func(t *testing.T) {
+		knockOnRoomWithStatus(t, knockingUser, roomID, "Can I knock anyways?", []string{"hs1"}, 403)
+	})
+
+	t.Run("Change the join rule of a room from 'invite' to 'xyz.amorgan.knock'", func(t *testing.T) {
+		inRoomUser.MustDo(
+			t,
+			"PUT",
+			[]string{"_matrix", "client", "r0", "rooms", roomID, "state", "m.room.join_rules", ""},
+			struct {
+				JoinRule string `json:"join_rule"`
+			}{
+				"xyz.amorgan.knock",
+			},
+		)
+	})
+
+	t.Run("Attempting to join a room with join rule 'xyz.amorgan.knock' without an invite should fail", func(t *testing.T) {
+		// Set server_name so we can find rooms via ID over federation
+		query := url.Values{
+			"server_name": []string{"hs1"},
+		}
+
+		knockingUser.MustDoWithStatusRaw(
+			t,
+			"POST",
+			[]string{"_matrix", "client", "r0", "join", roomID},
+			[]byte("{}"),
+			"application/json",
+			query,
+			403,
+		)
+	})
+
+	t.Run("Knocking on a room with join rule 'xyz.amorgan.xyz' should succeed", func(t *testing.T) {
+		knockOnRoom(t, knockingUser, roomID, testKnockReason, []string{"hs1"})
+	})
+
+	t.Run("A user that has already knocked cannot immediately knock again on the same room", func(t *testing.T) {
+		knockOnRoomWithStatus(t, knockingUser, roomID, "I really like knock knock jokes", []string{"hs1"}, 403)
+	})
+
+	// Now that Bob's knocked on the room successfully, let's check that everything went right
+
+	t.Run("parallel", func(t *testing.T) {
+		t.Run("Users in the room see a user's membership update when they knock", func(t *testing.T) {
+			t.Parallel()
+			inRoomUser.SyncUntilTimelineHas(t, roomID, func(ev gjson.Result) bool {
+				if ev.Get("type").Str != "m.room.member" || ev.Get("sender").Str != knockingUserUserID {
+					return false
+				}
+				must.EqualStr(t, ev.Get("content").Get("reason").Str, testKnockReason, "incorrect reason for knock")
+				must.EqualStr(t, ev.Get("content").Get("membership").Str, "xyz.amorgan.knock", "incorrect membership for knocking user")
+				return true
+			})
+		})
+
+		t.Run("Users see state events from the room that they knocked on in /sync", func(t *testing.T) {
+			t.Parallel()
+			knockingUser.SyncUntil(
+				t,
+				"",
+				"rooms."+client.GjsonEscape("xyz.amorgan.knock")+"."+client.GjsonEscape(roomID)+".knock_state.events",
+				func(ev gjson.Result) bool {
+					// We don't current define any required state event types to be sent
+					// If we've reached this point, then an entry for this room was found
+					return true
+				},
+			)
+		})
+	})
+
+	if !federation {
+		// Rescinding a knock over federation is currently not supported in Synapse
```

Comment:
Isn't it not even defined in the spec?

Reply:
*Rejecting* a knock over federation is currently not possible, but rescinding a knock over federation can work (implementations just need to figure out which homeserver to send the leave event through).

https://github.com/matrix-org/matrix-doc/pull/2403/files#diff-9dcacf9adb620c5346eba2f70f71bd50ed4f85a5e2844ac130e112165ce35c06R118-R119